README.md
<img src="https://user-images.githubusercontent.com/36012868/130392291-52b82b9b-fd52-424b-ba5a-b7630e9cf343.png" data-canonical-src="https://user-images.githubusercontent.com/36012868/130392291-52b82b9b-fd52-424b-ba5a-b7630e9cf343.png" height="200" width=400/>

[![English](https://img.shields.io/badge/Language-Russian-blue?style=plastic)](README.ru.md)

# Library for creating and customizing sliders

## Widget customization variations

Here are some examples of the library's features:

![](.github/switchers_example.gif)

## First I suggest you get familiar with what container and sliders are in the library

![](.github/container_sliders.png)

The library lets you change the style of containers and sliders using the fields in the table below

## List and description of the functionality:

| Fields           | Description                                   |
| ---------------- |:---------------------------------------------:|
| children     | Widgets that placed inside sliders / required |
| onSelect | The function takes an index variable that changes depending on the index of the current slider / required|
| containerHeight | Container height / required |
| containerWight | Container width / required |
| slidersColors | The color of each slider. A single color in the array will fill all sliders / white by default |
| slidersGradients | The gradient of each slider. A single gradient in the array will fill all sliders |
| containerBorder | Container border widget |
| slidersBorder | Slider border widget |
| containerBorderRadius | Container corner rounding radius |
| containerColor | Container fill color / grey by default |
| indents | Indents between the container and sliders (the same on all sides) / 0 by default |
| direction             | Defining the direction of the slider swipe / horizontal by default|
| isAllContainerTap     | A shadow cast by a box |
| containerBoxShadow    | Ability to tap on the current slider and change its index to the opposite (available only for 2 children) / false by default |
| initialIndex     | Initial index of the slider, which is rendering for the first time |

## Creating your own SlideSwitcher

SlideSwitcher is an ordinary widget. In order to create it you have to specify the obligatory parameters:
"children" - texts, icons, pictures and any other widgets that will be stored in the sliders;
"onSelect" - the function which will be executed during the change of the focused slider;
"containerHeight" - container height; "containerWight" - container width. To customize the widget
you should use the optional parameters from the table above.

#### Minimum functionality:

```
SlideSwitcher(
  children: [
    Text('First'),
    Text('Second'),
  ],
  onSelect: (index) {},
  containerHeight: 40,
  containerWight: 350,
),
```

To change the state of the screen it is recommended to create a variable storing the index of the current slider and call its setState
in the "onSelect" function

```
int switcherIndex1 = 0;

Column(
  children: [
    SlideSwitcher(
      children: [
        Text('First'),
        Text('Second'),
      ],
      onSelect: (index) => setState(() => switcherIndex1 = index),
      containerHeight: 40,
      containerWight: 350,
    ),
    const SizedBox(height: 20),
    if (switcherIndex1 == 0) ...[
      Container(height: 100, width: 100, color: Colors.red,)
    ]
    else ...[
      Container(height: 100, width: 100, color: Colors.green,)
    ],
  ],
),
```

## List of exceptions for cases that are not supported by the library:

| List of exceptions |
| :------------|
| All widgets from the list of "children" do not fit into the given container size |
| The "isAllContainerTap" parameter can be "true" only when "children" length is 2 |

*The first error from the table occurs when the smallest side of the container*
***(determined by the direction of the swipe)***
*multiplied by the length of the "children" by half is greater than the largest side of the container.*
*Only occurs when the borders of the container are maximum rounded*

# NEW
## Vertical Switchers.

![](.github/vertical_switchers.png)

The direction parameter of this library is responsible for the direction of the switch's movement.
Just add ***direction: Axis.vertical*** to your widget:

```
SlideSwitcher(
  children: [
    Text('First'),
    Text('Second'),
  ],
  onSelect: (index) {},
  containerHeight: 350,
  containerWight: 40,
  direction: Axis.vertical
),
```

## Empty Switchers

![](.github/switch_on.png)

If you need a switch with two children to determine whether to turn something on or off, then
add empty containers to children parameters and parameter ***isAllContainerTap: true***:

```
SlideSwitcher(
  children: [
    Container(),
    Container(),
  ],
  onSelect: (index) {},
  isAllContainerTap: true,
  containerHeight: 30,
  containerWight: 60,
),
```

[Google form for wishes and suggestions for the package](https://forms.gle/3Hghayy4yTnj1mjt7)


CHANGELOG.md
## 1.1.2

* Adding initial index parameter

## 1.1.1

* Adding exception handling for cases that are not supported by the library
* Added the ability to click on the whole container if it contains 2 children (isAllContainerTap parameter)
* Editing the exceptions of incorrect sliders size calculation

## 1.1.0

* Adding the ability to make vertical sliders
* Adding the ability to impose a shadow on the container
* Editing the container size error

## 1.0.0

* The first version of the user-friendly work of the package

## 0.0.1

* INITIAL RELEASE. Creating horizontal slide-switchers.


lib/slide_switcher.dart
library slide_switcher;

import 'package:flutter/material.dart';

class SlideSwitcher extends StatefulWidget {
  ///Widgets that placed inside sliders
  final List<Widget> children;

  ///The function takes an index variable, which varies depending on the index of the current slider
  final void Function(int index) onSelect;

  ///Container height
  final double containerHeight;

  ///Container width
  final double containerWight;

  ///The color of each slider. A single color in the array will fill all sliders
  final List<Color> slidersColors;

  ///The gradient of each slider. A single gradient in the array will fill all sliders
  final List<LinearGradient> slidersGradients;

  ///Container border widget
  final Border containerBorder;

  ///Slider border widget
  final Border slidersBorder;

  ///Container corner rounding radius
  final double containerBorderRadius;

  ///Container fill color
  final Color containerColor;

  ///Indents between the container and the sliders (the same on all sides)
  final double indents;

  ///Defining the direction of the slider swipe
  final Axis direction;

  ///A shadow cast by a box
  final List<BoxShadow> containerBoxShadow;

  ///Ability to tap on the current slider and change its index to the opposite (available only for 2 children)
  final bool isAllContainerTap;

  ///Initial index of the slider, which is rendering for the first time
  final int initialIndex;

  ///A class for creating sliders
  const SlideSwitcher({
    Key? key,
    required this.children,
    required this.containerHeight,
    required this.containerWight,
    required this.onSelect,
    this.containerBorder = const Border(),
    this.slidersBorder = const Border(),
    this.indents = 0,
    this.slidersColors = const [Colors.white],
    this.slidersGradients = const [],
    this.containerColor = Colors.grey,
    this.containerBorderRadius = 1000,
    this.direction = Axis.horizontal,
    this.containerBoxShadow = const [],
    this.isAllContainerTap = false,
    this.initialIndex = 0,
  }) : super(key: key);

  @override
  _SlideSwitcherState createState() => _SlideSwitcherState();
}

class _SlideSwitcherState extends State<SlideSwitcher>
    with SingleTickerProviderStateMixin {
  late final double sliderBorderRadius;
  late final double slidersHeight;
  late final double slidersWight;
  late final double containerBorderHeight;
  late final double containerBorderWight;
  late final bool verticalBadSize;
  late final bool horizontalBadSize;
  late final EdgeInsets padding;
  late int index;
  int lastIndex = -1;

  late final AnimationController _controller = AnimationController(
    duration: const Duration(milliseconds: 250),
    vsync: this,
  );
  late Animation<Offset> offsetAnimation = Tween<Offset>(
    begin: Offset(widget.initialIndex.toDouble(), 0.0),
    end: widget.direction == Axis.horizontal
        ? const Offset(1.0, 0.0)
        : const Offset(0.0, 1.0),
  ).animate(
    CurvedAnimation(
      parent: _controller,
      curve: Curves.linear,
    ),
  );

  @override
  void initState() {
    verticalBadSize = widget.direction == Axis.vertical &&
        widget.containerHeight < widget.containerWight * widget.children.length;

    horizontalBadSize = widget.direction == Axis.horizontal &&
        widget.containerHeight * widget.children.length > widget.containerWight;

    containerBorderHeight =
        widget.containerBorder.top.width + widget.containerBorder.bottom.width;

    containerBorderWight =
        widget.containerBorder.left.width + widget.containerBorder.right.width;

    sliderBorderRadius =
        (widget.containerHeight - widget.indents - containerBorderWight) *
            (widget.containerBorderRadius / widget.containerHeight);

    if (widget.direction == Axis.horizontal) {
      slidersHeight =
          widget.containerHeight - widget.indents * 2 - containerBorderHeight;
      padding = EdgeInsets.symmetric(horizontal: widget.indents);
      slidersWight =
          (widget.containerWight - widget.indents * 2 - containerBorderWight) /
              widget.children.length;
    } else {
      slidersWight =
          widget.containerWight - widget.indents * 2 - containerBorderWight;
      padding = EdgeInsets.symmetric(vertical: widget.indents);
      slidersHeight = (widget.containerHeight -
              widget.indents * 2 -
              containerBorderHeight) /
          widget.children.length;
    }
    index = widget.initialIndex;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    if (lastIndex == -1) _checkForExceptions();
    return Container(
      height: widget.children.isEmpty ? 0 : widget.containerHeight,
      width: widget.children.isEmpty ? 0 : widget.containerWight,
      decoration: BoxDecoration(
        color: widget.containerColor,
        borderRadius: BorderRadius.circular(widget.containerBorderRadius),
        border: widget.containerBorder,
        boxShadow: widget.containerBoxShadow,
      ),
      child: Stack(
        alignment: widget.direction == Axis.horizontal
            ? Alignment.centerLeft
            : Alignment.topCenter,
        children: [
          Padding(
            padding: padding,
            child: SlideTransition(
              position: offsetAnimation,
              child: AnimatedContainer(
                duration: const Duration(milliseconds: 250),
                height: slidersHeight,
                width: slidersWight,
                decoration: BoxDecoration(
                  border: widget.slidersBorder,
                  borderRadius: _makingBorder(index),
                  color: widget.slidersGradients.isEmpty
                      ? widget.slidersColors[
                          index + 1 > widget.slidersGradients.length
                              ? 0
                              : index]
                      : null,
                  gradient: widget.slidersGradients.isNotEmpty
                      ? widget.slidersGradients[
                          index + 1 > widget.slidersGradients.length
                              ? 0
                              : index]
                      : null,
                ),
              ),
            ),
          ),
          Padding(
            padding: padding,
            child: Flex(
              direction: widget.direction,
              children: List.generate(
                widget.children.length,
                (slidersIndex) => GestureDetector(
                  behavior: HitTestBehavior.translucent,
                  onTap: () {
                    _slidersOnTap(
                      widget.isAllContainerTap && widget.children.length == 2
                          ? 1 - index
                          : slidersIndex,
                    );
                  },
                  onHorizontalDragEnd: widget.direction == Axis.horizontal
                      ? (details) {
                          _slidersOnSwipe(details);
                        }
                      : null,
                  onVerticalDragEnd: widget.direction == Axis.vertical
                      ? (details) {
                          _slidersOnSwipe(details);
                        }
                      : null,
                  child: SizedBox(
                    height: slidersHeight,
                    width: slidersWight,
                    child: SizedBox(
                      child: Center(
                        child: widget.children[slidersIndex],
                      ),
                    ),
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }

  BorderRadius _makingBorder(int index) {
    return BorderRadius.only(
      bottomLeft: index == 0 && verticalBadSize ||
              index == widget.children.length - 1 && horizontalBadSize
          ? Radius.zero
          : Radius.circular(sliderBorderRadius),
      bottomRight: index == 0 && (verticalBadSize || horizontalBadSize)
          ? Radius.zero
          : Radius.circular(sliderBorderRadius),
      topLeft: index == widget.children.length - 1 && verticalBadSize ||
              index == widget.children.length - 1 && horizontalBadSize
          ? Radius.zero
          : Radius.circular(sliderBorderRadius),
      topRight: index == widget.children.length - 1 && verticalBadSize ||
              index == 0 && horizontalBadSize
          ? Radius.zero
          : Radius.circular(sliderBorderRadius),
    );
  }

  void _slidersOnTap(int slidersIndex) {
    lastIndex = index;
    index = slidersIndex;
    setState(() {});
    widget.onSelect(index);
    if (widget.children.length == 2) {
      index == 1 ? _controller.forward() : _controller.reverse();
    } else {
      if (widget.direction == Axis.horizontal) {
        offsetAnimation = Tween<Offset>(
          begin: Offset(lastIndex.toDouble(), 0.0),
          end: Offset(index.toDouble(), 0.0),
        ).animate(
          CurvedAnimation(
            parent: _controller,
            curve: Curves.linear,
          ),
        );
      } else {
        offsetAnimation = Tween<Offset>(
          begin: Offset(0.0, lastIndex.toDouble()),
          end: Offset(0.0, index.toDouble()),
        ).animate(
          CurvedAnimation(
            parent: _controller,
            curve: Curves.linear,
          ),
        );
      }
      _controller.reset();
      _controller.forward();
    }
  }

  void _slidersOnSwipe(DragEndDetails details) {
    if (details.primaryVelocity! > 0 && index != widget.children.length - 1) {
      _slidersOnTap(index + 1);
    }
    if (details.primaryVelocity! < 0 && index != 0) {
      _slidersOnTap(index - 1);
    }
  }

  void _checkForExceptions() {
    if (widget.isAllContainerTap && widget.children.length != 2) {
      debugPrint(
          '\x1B[31mThe "isAllContainerTap" parameter can be "true" only when "children" length is 2\nRemove "isAllContainerTap" or make 2 "children"\x1B[0m');
      throw 'The "isAllContainerTap" parameter can be "true" only when "children" length is 2\nRemove "isAllContainerTap" or make 2 "children"';
    }

    double shortSide = widget.containerHeight > widget.containerWight
        ? widget.containerWight
        : widget.containerHeight;
    double longSide = widget.containerHeight > widget.containerWight
        ? widget.containerHeight
        : widget.containerWight;
    if (shortSide * widget.children.length * 0.5 > longSide &&
        widget.containerBorderRadius * 2 > shortSide &&
        widget.children.length != 2) {
      debugPrint(
          '\x1B[31mAll widgets from the list of "children" do not fit into the given container size.\nTry applying other container sizes\x1B[0m');
      throw 'All widgets from the list of "children" do not fit into the given container size.\nTry applying other container sizes';
    }
    lastIndex = 0;
  }
}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:slide_switcher/slide_switcher.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        scaffoldBackgroundColor: Colors.white10.withOpacity(0.27),
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int switcherIndex1 = 0;
  int switcherIndex2 = 0;
  int switcherIndex3 = 0;
  int switcherIndex4 = 0;
  int switcherIndex5 = 0;
  int switcherIndex6 = 0;
  int switcherIndex7 = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: switcherIndex1 == 0
          ? Colors.white10.withOpacity(0.27)
          : Colors.white10.withOpacity(0.2),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            SlideSwitcher(
              children: [
                Container(),
                Container(),
              ],
              containerHeight: 30,
              containerWight: 60,
              isAllContainerTap: true,
              onSelect: (int index) => setState(() => switcherIndex1 = index),
              indents: 5,
              containerColor:
                  switcherIndex1 == 0 ? Colors.grey : Colors.indigo.shade300,
            ),
            const SizedBox(height: 20),
            SlideSwitcher(
              children: [
                Text(
                  'First',
                  style: TextStyle(
                    fontWeight:
                        switcherIndex2 == 0 ? FontWeight.w500 : FontWeight.w400,
                    color:
                        switcherIndex2 == 0 ? Colors.deepOrange : Colors.white,
                  ),
                ),
                Text(
                  'Second',
                  style: TextStyle(
                    fontWeight:
                        switcherIndex2 == 1 ? FontWeight.w500 : FontWeight.w400,
                    color:
                        switcherIndex2 == 1 ? Colors.deepOrange : Colors.white,
                  ),
                ),
              ],
              containerColor: Colors.deepOrange,
              onSelect: (int index) => setState(() => switcherIndex2 = index),
              containerHeight: 40,
              containerWight: 350,
            ),
            const SizedBox(height: 20),
            SlideSwitcher(
              children: const [
                Text(
                  'First',
                  style: TextStyle(
                    fontWeight: FontWeight.w600,
                    color: Color(0xff714F43),
                  ),
                ),
                Text(
                  'Second',
                  style: TextStyle(
                    fontWeight: FontWeight.w600,
                    color: Color(0xff714F43),
                  ),
                ),
                Text(
                  'Third',
                  style: TextStyle(
                    fontWeight: FontWeight.w600,
                    color: Color(0xff714F43),
                  ),
                ),
                Text(
                  'Forth',
                  style: TextStyle(
                    fontWeight: FontWeight.w600,
                    color: Color(0xff714F43),
                  ),
                ),
              ],
              initialIndex: 2,
              onSelect: (int index) => setState(() => switcherIndex3 = index),
              containerBorderRadius: 0,
              indents: 10,
              containerBorder: Border.all(
                width: 3,
                color: const Color(0xffFFFFE3),
              ),
              containerColor: const Color(0xffE1CCB9).withOpacity(0.8),
              slidersColors: const [
                Color(0xffFFFFE3),
              ],
              containerHeight: 50,
              containerWight: 350,
            ),
            const SizedBox(height: 20),
            SlideSwitcher(
              children: [
                Text(
                  'First',
                  style: TextStyle(
                      fontSize: 15,
                      color: switcherIndex4 == 0
                          ? Colors.white.withOpacity(0.9)
                          : Colors.grey),
                ),
                Text(
                  'Second',
                  style: TextStyle(
                      fontSize: 15,
                      color: switcherIndex4 == 1
                          ? Colors.white.withOpacity(0.9)
                          : Colors.grey),
                ),
                Text(
                  'Third',
                  style: TextStyle(
                      fontSize: 15,
                      color: switcherIndex4 == 2
                          ? Colors.white.withOpacity(0.9)
                          : Colors.grey),
                ),
              ],
              onSelect: (int index) => setState(() => switcherIndex4 = index),
              containerColor: Colors.transparent,
              containerBorder: Border.all(color: Colors.white),
              slidersGradients: const [
                LinearGradient(
                  colors: [
                    Color.fromRGBO(47, 105, 255, 1),
                    Color.fromRGBO(188, 47, 255, 1),
                  ],
                ),
                LinearGradient(
                  colors: [
                    Color.fromRGBO(47, 105, 255, 1),
                    Color.fromRGBO(0, 192, 169, 1),
                  ],
                ),
                LinearGradient(
                  colors: [
                    Color.fromRGBO(255, 105, 105, 1),
                    Color.fromRGBO(255, 62, 62, 1),
                  ],
                ),
              ],
              indents: 9,
              containerHeight: 50,
              containerWight: 315,
            ),
            const SizedBox(height: 20),
            Row(
              children: [
                const SizedBox(width: 65),
                SlideSwitcher(
                  children: const [
                    Icon(
                      Icons.account_circle_rounded,
                      color: Colors.white,
                    ),
                    Icon(
                      Icons.add_moderator,
                      color: Colors.white,
                    ),
                  ],
                  onSelect: (int index) =>
                      setState(() => switcherIndex5 = index),
                  direction: Axis.vertical,
                  containerColor: Colors.teal.withOpacity(0.5),
                  slidersColors: const [Colors.teal],
                  containerHeight: 70,
                  containerWight: 90,
                ),
                const SizedBox(width: 20),
                SlideSwitcher(
                  children: const [
                    Text(
                      'First',
                      style: TextStyle(
                        color: Color.fromRGBO(53, 90, 242, 1),
                        fontWeight: FontWeight.w500,
                      ),
                    ),
                    Text(
                      'Second',
                      style: TextStyle(
                        color: Color.fromRGBO(53, 90, 242, 1),
                        fontWeight: FontWeight.w500,
                      ),
                    ),
                    Text(
                      'Third',
                      style: TextStyle(
                        color: Color.fromRGBO(53, 90, 242, 1),
                        fontWeight: FontWeight.w500,
                      ),
                    ),
                    Text(
                      'Forth',
                      style: TextStyle(
                        color: Color.fromRGBO(53, 90, 242, 1),
                        fontWeight: FontWeight.w500,
                      ),
                    ),
                    Text(
                      'Fifth',
                      style: TextStyle(
                        color: Color.fromRGBO(53, 90, 242, 1),
                        fontWeight: FontWeight.w500,
                      ),
                    ),
                  ],
                  onSelect: (int index) =>
                      setState(() => switcherIndex6 = index),
                  slidersColors: const [Colors.white],
                  containerHeight: 320,
                  containerWight: 60,
                  indents: 4,
                  direction: Axis.vertical,
                  containerColor: const Color.fromRGBO(140, 176, 254, 1),
                ),
                const SizedBox(width: 20),
                SlideSwitcher(
                  children: const [
                    Text(
                      'F',
                      style: TextStyle(
                        color: Colors.white,
                        fontWeight: FontWeight.w500,
                      ),
                    ),
                    Text(
                      'S',
                      style: TextStyle(
                        color: Colors.white,
                        fontWeight: FontWeight.w500,
                      ),
                    ),
                  ],
                  direction: Axis.vertical,
                  containerColor: Colors.purple,
                  slidersColors: [Colors.transparent],
                  slidersBorder: Border.all(color: Colors.white, width: 2),
                  containerBorder: Border.all(color: Colors.white, width: 2),
                  containerHeight: 300,
                  containerWight: 40,
                  indents: 5,
                  onSelect: (int index) =>
                      setState(() => switcherIndex7 = index),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}
